tests/test_speed.py

```diff
@@ -137,190 +163,107 @@ async def test_class_creation_speed(self):
-                            arrays_read_speed * u.ct / u.second,
-                        )
-                    )
+            dt = time.monotonic() - t0
+            arrays_read_speed = num_samples / dt
+            nlost = data.int0[0] - num_samples
+            print(
+                f"Read {arrays_read_speed:0.0f} arrays samples/second "
+                f"({num_samples} samples); "
+                f"lost {nlost} samples"
+            )
 
-                await salobj.set_summary_state(
-                    remote=remote, state=salobj.State.STANDBY,
-                )
-                await salobj.set_summary_state(
-                    remote=remote,
-                    state=salobj.State.ENABLED,
-                    settingsToApply="logLevel",
+            if verify is not None:
+                self.insert_measurement(
+                    verify.Measurement(
+                        "salobj.ReadTest_arrays", arrays_read_speed * u.ct / u.second,
+                    )
                 )
 
-                # Wait for the first sample so we know the writer is running.
+            await salobj.set_summary_state(
+                remote=remote, state=salobj.State.STANDBY,
+            )
+            await salobj.set_summary_state(
+                remote=remote, state=salobj.State.ENABLED, settingsToApply="logLevel",
+            )
+
+            # Wait for the first logLevel sample. This is automatically
+            # output by the Controller and probably has level=20.
+            data = await remote.evt_logLevel.next(flush=False, timeout=STD_TIMEOUT)
+            self.assertGreater(data.level, 0)
+            # Wait for the next logLevel sample, which is the first
+            # one output by the write loop.
+            data = await remote.evt_logLevel.next(flush=False, timeout=STD_TIMEOUT)
+            self.assertEqual(data.level, 0)
+            t0 = time.monotonic()
+            for i in range(num_samples):
                 data = await remote.evt_logLevel.next(flush=False, timeout=STD_TIMEOUT)
-                t0 = time.monotonic()
-                for i in range(num_samples):
-                    data = await remote.evt_logLevel.next(
-                        flush=False, timeout=STD_TIMEOUT
-                    )
-                dt = time.monotonic() - t0
-                log_level_read_speed = num_samples / dt
-                print(f"read {log_level_read_speed:0.0f} logLevel samples/second")
-                self.assertTrue(data.level, num_samples)
-
-                if verify is not None:
-                    self.insert_measurement(
-                        verify.Measurement(
-                            "salobj.ReadTest_logLevel",
-                            log_level_read_speed * u.ct / u.second,
-                        )
-                    )
+            dt = time.monotonic() - t0
+            log_level_read_speed = num_samples / dt
+            nlost = data.level - num_samples
```

Comment:
Is data.level fixed or incremented during use above? I don't see anything that looks like it getting incremented above and I didn't think python had log levels higher than 50. This looks like I would always see nlost < 0. What am I missing?

Reply:
I am using the logLevel message as a representative "small topic" (one of smallest available to me). The data being sent is not a log level, but an incrementing integer.

Writing DDS samples is faster than reading, so with enough samples I would expect the read queues to overfill and messages to be lost. A symptom of that is that some of the incremented values will be lost, and so there will be gaps in the values read. For example, suppose the read values are 0, 1, 2, 3, 7, 8, 9 (in reality I would expect losses to start at much higher values). I measure samples lost as the difference between the final read value and the number of samples read; in this case 9 - 6 = 3 messages lost (the values 4, 5 and 6).